### ISSUE TITLE
Screen readers don't read radioButtons() labels

### ISSUE THREAD

**Issue by eliocamp** [rstudio/shiny#3105]
The radio buttons created with radioButtons() don't work well with screen readers. I tired both with Orca and ChromeVox and neither of them read their label. 

To see the difference, this code creates two groups of radio buttons. Launch the app in Chrome with ChromeVox installed and navigate between them with tab. 

The when navigating to the first one it speaks "selected normal radio button selected" which means that there is no context of that "normal" means. When navigating to the second one it speaks "Distribution type radio button group normal radio button selected", which lets the user hear the context of the radio button.

```r
aria_radio <-  function (inputId, label, choices = NULL, selected = NULL, inline = FALSE, 
                         width = NULL, choiceNames = NULL, choiceValues = NULL) {
  args <- shiny:::normalizeChoicesArgs(choices, choiceNames, choiceValues)
  selected <- shiny::restoreInput(id = inputId, default = selected)
  selected <- if (is.null(selected)) 
    args$choiceValues[[1]]
  else as.character(selected)
  if (length(selected) > 1) 
    stop("The 'selected' argument must be of length 1")
  
  options <- aria_generateOptions(inputId, selected, inline, "radio",
                                     args$choiceNames, args$choiceValues)
  divClass <- "form-group shiny-input-radiogroup shiny-input-container"
  if (inline) shiny::divClass <- paste(divClass, "shiny-input-container-inline")
  
  
  shiny::tags$div(id = inputId, style = if (!is.null(width)) paste0("width: ", shiny::validateCssUnit(width), ";"),
                  class = divClass, role = "radiogroup", 
                  `aria-labelledby` = paste0(inputId, "label"),                                   # this is new
                  shiny::tags$label(label, class = "control-label", class = if (is.null(label)) "shiny-label-null", 
                                    `for` = inputId, 
                                    id = paste0(inputId, "label")                                         # this is also new.
                                    ),
                  options)
}

aria_generateOptions <- function (inputId, selected, inline, type = "checkbox", choiceNames, 
                                  choiceValues, session = shiny::getDefaultReactiveDomain()) {
  options <- mapply(choiceValues, choiceNames, FUN = function(value, 
                                                              name) {
    inputTag <- shiny::tags$input(type = type, name = inputId, 
                                  value = value)
    if (value %in% selected) {
      inputTag$attribs$checked <- "checked"
    } else {
    }
    
    if (type == "radio") {
      inputTag$attribs$role <- "radio"        # This is also needed
    }
      
    pd <- shiny:::processDeps(name, session)
    if (inline) {
      shiny::tags$label(class = paste0(type, "-inline"), inputTag, 
                        shiny::tags$span(pd$html, pd$deps))
    }
    else {
      shiny::tags$div(class = type, shiny::tags$label(inputTag, shiny::tags$span(pd$html, pd$deps)))
    }
  }, SIMPLIFY = FALSE, USE.NAMES = FALSE)
  shiny::div(class = "shiny-options-group", options)
}


ui <- fluidPage(
  radioButtons("dist", "Distribution type:",
               c("Normal" = "norm",
                 "Uniform" = "unif",
                 "Log-normal" = "lnorm",
                 "Exponential" = "exp")),
  
  aria_radio("dist2", "Distribution type:",
               c("Normal" = "norm",
                 "Uniform" = "unif",
                 "Log-normal" = "lnorm",
                 "Exponential" = "exp"))
  
  ,
  plotOutput("distPlot")
)

server <- function(input, output) {
  output$distPlot <- renderPlot({
    dist <- switch(input$dist,
                   norm = rnorm,
                   unif = runif,
                   lnorm = rlnorm,
                   exp = rexp,
                   rnorm)
    
    hist(dist(500))
  })
}

shinyApp(ui, server)
```

I created the above `aria_radio()` by copying the example here: https://www.w3.org/TR/2017/WD-wai-aria-practices-1.1-20170628/examples/radio/radio-1/radio-1.html
I'm the furthest you can be to an expert in this matter, so I wouldn't trus that this code is a complete fix.

**Comment by cpsievert** [comment]
Have you tried out `radioButtons()` with development version of shiny (I think we recently solved this)? `remotes::install_github("rstudio/shiny")`

**Comment by eliocamp** [comment]
Seems to have the same behaviour :(
(I tried to record a screencap but I can't manage to record the sound coming from the computer)

**Comment by cpsievert** [comment]
@jooyoungseo any idea why @eliocamp might be having issues with their screen reader? Do we need the additional `role`/`aria-labelledby` that they've suggested here?

**Comment by jooyoungseo** [comment]
Hello @eliocamp. Thank you for filing the issue. I was working for accessibility on Shiny team this summer (I am an actual screen reader user).

@cpsievert, I promise that I will take a close look at this issue by this weekend if this is not a rush issue.

**Comment by eliocamp** [comment]
Great! I'm using that snippet of code as a workaround and on my end it seems to be working OK for my specific case, so no rush at all.

**Comment by jooyoungseo** [comment]
@eliocamp, this is a nice call!

@cpsievert, I have pushed a PR to address this issue with a simpler code.

**Comment by eliocamp** [comment]
Hi! I came into the same problem with checkboxGroupInput. ChromeVox doesn't read the label.

**Comment by jooyoungseo** [comment]
I had already addressed that issue via PR #3110.

**Comment by eliocamp** [comment]
Ah, my bad. I thought it only fixed it for radiobuttons.

**Comment by wch** [comment]
This is superseded by #3194 since I'm not able to push to this branch.

**Comment by cpsievert** [comment]
For posterity, I believe this issue was actually closed by #3187 (not #3194)
